### PR TITLE
Refactor completion variants sorting: use weighers

### PR DESCRIPTION
**Pull request by vlad20012**
This PR changes nothing for the end-user (i.e. the new sorting should work exactly like the old one).
For us, this tidies up completion variants sorting. Please read docs for `RS_COMPLETION_WEIGHERS`

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/completion/sort/RsCompletionSorting.kt [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        if (element is RsLookupElement) property(element.props) else false
+
+    override val id: String get() = id
+}
+
+private fun preferUpperVariant(
+    property: (P) -> Enum<*>,
+    id: String
+): RsCompletionWeigher = object : RsCompletionWeigher {
+    // `0` is sorted before `1`
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/completion/sort/RsCompletionSorting.kt:136]: [minor]
```suggestion
    // `0` is placed before `1`
```

File: src/main/kotlin/org/rust/lang/core/completion/sort/RsCompletionWeigher.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import com.intellij.codeInsight.lookup.LookupElement
+
+/**
+ * A weigher for Rust completion variants.
+ *
+ * @see RS_COMPLETION_WEIGHERS
+ */
+interface RsCompletionWeigher {
+    /**
+     * Returned values will be sorted in ascending order, i.e.
```
Comment (ortem) [inline src/main/kotlin/org/rust/lang/core/completion/sort/RsCompletionWeigher.kt:17]: [minor]
```suggestion
     * Returned values are sorted in ascending order, i.e.
```

**Review by ortem — APPROVED** [review]
Generally LGTM

**Comment by vlad20012** [comment]
bors r=ortem

**Pull request merged**